ta/pkcs11/src/processing_aes.c

```diff
@@ -14,6 +15,385 @@
+ * buffer at AE finalization.
+ *
+ * As per PKCS#11, CCM/GCM decrypt expect the tag/mac data to be provided
+ * inside the input data for DecryptUpdate() and friends. But the DecryptFinal
+ * API does not provide input data reference hence we do not know which is the
+ * last call to DecryptUpdate() where last bytes are not ciphered data but the
+ * requested tag/mac byte. To handle this, the TA saves the last input data
+ * bytes (length is defined by the tag byte size) in the AE context and
+ * waits the DecryptFinal() to either treat these as data bytes or tag/mac
+ * bytes. Refer to pending_tag and pending_size in struct ae_aes_context.
+ */
+
+/*
+ * @size - byte size of the allocated buffer
+ * @data - pointer to allocated data
+ */
+struct out_data_ref {
+	size_t size;
+	void *data;
+};
+
+/*
+ * @tag_byte_len - tag size in byte
+ * @pending_tag - Input data that could be the appended tag
+ * @pending_size - Size of pending input data that could be the tag
+ * @out_data - Pointer to an array of output data references.
+ * @out_count - Number of buffer references in out_data
+ */
+struct ae_aes_context {
+	size_t tag_byte_len;
+	char *pending_tag;
+	size_t pending_size;
+	struct out_data_ref *out_data;
+	size_t out_count;
+};
+
+static void release_ae_aes_context(struct ae_aes_context *ctx)
+{
+	size_t n = 0;
+
+	for (n = 0; n < ctx->out_count; n++)
+		TEE_Free(ctx->out_data[n].data);
+
+	TEE_Free(ctx->out_data);
+	ctx->out_data = NULL;
+	ctx->out_count = 0;
+
+	TEE_Free(ctx->pending_tag);
+	ctx->pending_tag = NULL;
+}
+
+enum pkcs11_rc tee_ae_decrypt_update(struct pkcs11_session *session,
+				     void *in, size_t in_size)
+{
+	struct ae_aes_context *ctx = session->processing->extra_ctx;
+	size_t data_len = 0;
+	size_t size = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	enum pkcs11_rc rc = 0;
+	char *ct = NULL;
+	size_t ct_size = 0;
+	void *ptr = NULL;
+
+	if (!in_size)
+		return PKCS11_CKR_OK;
+
+	if (!in)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	/*
+	 * Save the last input bytes in case they are the tag
+	 * bytes and not ciphered data bytes to be decrypted.
+	 */
+
+	if (ctx->pending_size + in_size <= ctx->tag_byte_len) {
+		/*
+		 * Data bytes are all potential tag bytes.
+		 * We only need to update the pending_tag buffer,
+		 * and cannot treat any byte as data byte.
+		 */
+		TEE_MemMove(ctx->pending_tag + ctx->pending_size, in, in_size);
+
+		ctx->pending_size += in_size;
+
+		return PKCS11_CKR_OK;
+	}
+
+	/* Size of data that are not potential tag in pending and input data */
+	data_len = in_size + ctx->pending_size - ctx->tag_byte_len;
+
+	if (ctx->pending_size &&
+	    (ctx->pending_size + in_size) >= ctx->tag_byte_len) {
+		/* Process pending tag bytes that are effective data byte */
+		uint32_t len = MIN(data_len, ctx->pending_size);
+
+		res = TEE_AEUpdate(session->processing->tee_op_handle,
+				   ctx->pending_tag, len, NULL, &ct_size);
+
+		// TODO: explain this
```

Comment:
Let's explain this!

Reply:
It should be somehow related to to my concern above https://github.com/OP-TEE/optee_os/pull/6467#issuecomment-1824629664